### PR TITLE
Add Cache-Control: no-store to API responses

### PR DESCRIPTION
**Pull request by nawatts**
During production readiness review, many services encounter a security finding related to lack of cache control headers.

### CODE REVIEW

**Review by aherbst-broad — APPROVED** [review]
Thanks for doing this.

**Review by pshapiro4broad — APPROVED** [review]
Just a bunch of small nits, this seems OK overall.

Although I wonder about making this behavior the default for all APIs. Aren't there some responses that it's OK for a client to cache? When I've done this before, it's usually a matter of marking specific responses as no-cache. If there's Jira ticket (maybe from the incident) can it be linked in the PR description?

**Comment by nawatts** [comment]
Setting this to draft until I can incorporate some feedback from the corresponding WDS PR: https://github.com/DataBiosphere/terra-workspace-data-service/pull/779

File: service/src/main/java/bio/terra/javatemplate/controller/ControllerConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class ControllerConfiguration {
+  /**
+   * API responses contain potentially sensitive data and thus should not be cached. This configures
+   * a filter that adds a "Cache-Control: no-store" header to API responses.
+   */
+  @Bean
```
[the remark this answers is not in the document]
Comment (nawatts) [reply service/src/main/java/bio/terra/javatemplate/controller/ControllerConfiguration.java]: I was blindly following WDS' configuration for a filter that sets headers.

File: service/src/main/java/bio/terra/javatemplate/controller/ControllerConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  public FilterRegistrationBean<CacheControlNoStoreFilter> cacheControlNoStoreFilter() {
+    FilterRegistrationBean<CacheControlNoStoreFilter> registrationBean =
+        new FilterRegistrationBean<>();
+    registrationBean.setFilter(new CacheControlNoStoreFilter());
```
[the remark this answers is not in the document]
Comment (nawatts) [reply service/src/main/java/bio/terra/javatemplate/controller/ControllerConfiguration.java]: Ooh, today I learned!

File: service/src/main/java/bio/terra/javatemplate/controller/ControllerConfiguration.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (nawatts) [reply service/src/main/java/bio/terra/javatemplate/controller/ControllerConfiguration.java]: Resolved with https://github.com/DataBiosphere/terra-java-project-template/pull/30/commits/ef10b774df6848471815921e6b57df4f6993b57c.

File: service/src/main/java/bio/terra/javatemplate/controller/ControllerConfiguration.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (nawatts) [reply service/src/main/java/bio/terra/javatemplate/controller/ControllerConfiguration.java]: Resolved with https://github.com/DataBiosphere/terra-java-project-template/pull/30/commits/ef10b774df6848471815921e6b57df4f6993b57c.

File: service/src/main/java/bio/terra/javatemplate/service/CacheControlNoStoreFilter.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package bio.terra.javatemplate.service;
+
+import jakarta.servlet.*;
```
[the remark this answers is not in the document]
Comment (nawatts) [reply service/src/main/java/bio/terra/javatemplate/service/CacheControlNoStoreFilter.java]: Resolved with https://github.com/DataBiosphere/terra-java-project-template/pull/30/commits/ef10b774df6848471815921e6b57df4f6993b57c.

File: service/src/main/java/bio/terra/javatemplate/controller/ControllerConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+@Configuration
+public class ControllerConfiguration {
+  /**
+   * API responses contain potentially sensitive data and thus should not be cached. This configures
+   * a filter that adds a "Cache-Control: no-store" header to API responses.
+   */
+  @Bean
+  public FilterRegistrationBean<CacheControlNoStoreFilter> cacheControlNoStoreFilter() {
+    FilterRegistrationBean<CacheControlNoStoreFilter> registrationBean =
+        new FilterRegistrationBean<>();
```
[the remark this answers is not in the document]
Comment (nawatts) [reply service/src/main/java/bio/terra/javatemplate/controller/ControllerConfiguration.java]: Resolved with https://github.com/DataBiosphere/terra-java-project-template/pull/30/commits/ef10b774df6848471815921e6b57df4f6993b57c.

File: service/src/main/java/bio/terra/javatemplate/service/CacheControlNoStoreFilter.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+import jakarta.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class CacheControlNoStoreFilter implements Filter {
+  @Override
+  public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+      throws IOException, ServletException {
+    var servletResponse = (HttpServletResponse) response;
+    servletResponse.setHeader("Cache-Control", "no-store");
```
[the remark this answers is not in the document]
Comment (nawatts) [reply service/src/main/java/bio/terra/javatemplate/service/CacheControlNoStoreFilter.java]: Switched to the Spring constant in https://github.com/DataBiosphere/terra-java-project-template/pull/30/commits/e591247de829e8574639a989dd303845737cd799.

**Comment by nawatts** [comment]
I also just noticed this `Cache-Control` configuration in `application.yml`: https://github.com/DataBiosphere/terra-java-project-template/blob/fbc67befdf77be0630f2dd0ff73991907f4dd24a/service/src/main/resources/application.yml#L51-L55

I'm inclined to remove it because:
- The security finding specifically mentions `no-store`, where this configuration sets `max-age=0`.
- It also applies to static resources, which ideally could be cached.

**Review by pshapiro4broad — APPROVED** [review]
Looks good to me. Not sure about how this interacts with the application.yml config; the fact that it can also be configured in application properties means that it could be overridden in a helm config too, I wonder if that's happening at all.

**Review by aherbst-broad — APPROVED** [review]

**Comment by nawatts** [comment]
Since the project template already includes configuration for cache control (https://github.com/DataBiosphere/terra-java-project-template/pull/30#issuecomment-2104772929), I'll err on the side of "no change" and close this.

**Pull request closed without merging**